tsdb/chunks/head_chunks_test.go

```diff
@@ -0,0 +1,301 @@
+)
+
+func TestHeadReadWriter_WriteChunk(t *testing.T) {
+	hrw, close := testHeadReadWriter(t)
+	defer func() {
+		testutil.Ok(t, hrw.Close())
+		close()
+	}()
+
+	expectedBytes := []byte{}
+	numChunks := 2000
+	nextChunkOffset := uint64(HeadChunkFileHeaderSize + SeriesRefSize + 2*MintMaxtSize)
+	chkCRC32 := newCRC32()
+
+	var buf [8]byte
+	for i := 0; i < numChunks; i++ {
+		seriesRef := uint64(rand.Int63())
+		mint := int64(i*1000 + 1)
+		maxt := int64((i + 1) * 1000)
+		chunk := randomChunk(t)
+		chkCRC32.Reset()
+
+		chkRef, err := hrw.WriteChunk(seriesRef, mint, maxt, chunk)
+		testutil.Ok(t, err)
+		testutil.Equals(t, chunkRef(1, nextChunkOffset), chkRef)
+
+		// Calculating expected bytes written on disk.
+		binary.BigEndian.PutUint64(buf[:], seriesRef)
+		expectedBytes = append(expectedBytes, buf[:SeriesRefSize]...)
+		_, err = chkCRC32.Write(buf[:SeriesRefSize])
+		testutil.Ok(t, err)
+
+		binary.BigEndian.PutUint64(buf[:], uint64(mint))
+		expectedBytes = append(expectedBytes, buf[:MintMaxtSize]...)
+		_, err = chkCRC32.Write(buf[:MintMaxtSize])
+		testutil.Ok(t, err)
+
+		binary.BigEndian.PutUint64(buf[:], uint64(maxt))
+		expectedBytes = append(expectedBytes, buf[:MintMaxtSize]...)
+		_, err = chkCRC32.Write(buf[:MintMaxtSize])
+		testutil.Ok(t, err)
+
+		expectedBytes = append(expectedBytes, byte(chunk.Encoding()))
+		_, err = chkCRC32.Write([]byte{byte(chunk.Encoding())})
+		testutil.Ok(t, err)
+
+		n := binary.PutUvarint(buf[:], uint64(len(chunk.Bytes())))
+		expectedBytes = append(expectedBytes, buf[:n]...)
+		_, err = chkCRC32.Write(buf[:n])
+		testutil.Ok(t, err)
+
+		expectedBytes = append(expectedBytes, chunk.Bytes()...)
+		_, err = chkCRC32.Write(chunk.Bytes())
+		testutil.Ok(t, err)
+
+		expectedBytes = append(expectedBytes, chkCRC32.Sum(nil)...)
+
+		// += encoding + chunk data len + chunk data + CRC + seriesRef,mint,maxt of next chunk.
+		nextChunkOffset += 1 + uint64(n) + uint64(len(chunk.Bytes())) + CRCSize + SeriesRefSize + 2*MintMaxtSize
+	}
+	testutil.Ok(t, hrw.flushBuffer())
+
+	testutil.Assert(t, len(hrw.mmappedChunkFiles) == 1 && len(hrw.closers) == 1, "expected only 1 mmapped file, got %d", len(hrw.mmappedChunkFiles))
+
+	fileName := hrw.curFile.Name()
+	actualBytes, err := ioutil.ReadFile(fileName)
+	testutil.Ok(t, err)
+
+	// Check header of the segment file.
+	testutil.Equals(t, MagicHeadChunks, int(binary.BigEndian.Uint32(actualBytes[0:MagicChunksSize])))
+	testutil.Equals(t, chunksFormatV1, int(actualBytes[MagicChunksSize]))
+
+	mint := binary.BigEndian.Uint64(actualBytes[HeaderMintOffset : HeaderMintOffset+8])
+	maxt := binary.BigEndian.Uint64(actualBytes[HeaderMaxtOffset : HeaderMaxtOffset+8])
+	testutil.Assert(t, mint > 0, "expected mint >0")
+	testutil.Assert(t, maxt == 0, "expected maxt = 0 for an active file")
+
+	// Remaining chunk data.
+	testutil.Equals(t, expectedBytes, actualBytes[HeadChunkFileHeaderSize:])
+}
+
+func TestHeadReadWriter_ReadChunk(t *testing.T) {
```

Comment:
I still think that you can combine this test with the previouse one, the only dif I see that you just check the actual chunk content so this should just add no more than few lines to the test above, no?

Reply:
Another difference is that in WriteChunk I am only filling up 1 file and checking it's contents. In ReadChunk, I am having >1 files. So while combining those tests is possible, the simplicity of reading the test will go away by trying to handle that case.

And I feel explicitly splitting the reading and writing tests would make it easier to find bugs if any. Open for opinions on this.

Reply:
yep, I don't mind, maybe if you put the boostraping for the data in a separate func and reuse it in all tests?

Reply:
ok I just had another look and yes one test is named read and the other is named write, but both write the chunk to disk one both read it after that and theck different bits, so unless combining them would make it very very hard to follow I would vote to have it as a single one.